Logical Replication How-To (#303)
crate/crate-howtos: docs/clustering/logical-replication-setup.rst

.. _logical_replication_setup:

==================================================
Logical replication setup between CrateDB clusters
==================================================

Logical replication is a method of data replication between multiple clusters.
As a publish/subscribe model, it allows a publishing cluster to make certain
tables available for subscription. Subscribing clusters subscribe to a

> publication to replay any changes to the data on their side.

Reviewer: The last sentence is a bit complicate IMO. Maybe "Any change on a published table will be replicated to the matching table in the subscribing cluster"?

  Logical replication is available in CrateDB 4.8 and later.

To replicate data from one cluster to another, we first of all need two CrateDB
clusters. For this example, we spin them up locally as two single-node Docker
containers:

.. code-block:: console

  sh$ docker run \
        --name cluster-1 \
        --detach \
        --publish 4201:4200 \
        --publish 5433:5432 \
        crate/crate:nightly-4.8.0-20220401 \
        -Cdiscovery.type=single-node
  sh$ docker run \
        --name cluster-2 \
        --detach \
        --publish 4202:4200 \
        --publish 5434:5432 \
        crate/crate:nightly-4.8.0-20220401 \
        -Cdiscovery.type=single-node

We apply an offset to the ports of each cluster, so they don't overlap. The
Admin UIs of ``cluster-1`` and ``cluster-2`` are accessible at
http://localhost:4201 and http://localhost:4202 respectively. All SQL
statements discussed below can be executed via the corresponding Admin UI.

There exists no fixed designation of a cluster being a subscriber or publisher.
One cluster can even be both for different tables. Nevertheless, we will stick
for this tutorial to ``cluster-1`` taking the role of the publishing cluster

> (data source). ``cluster-2`` will subscribe to ``cluster-1`` (data sink).

Reviewer: I think a more telling name for the two clusters would be helpful further down in the article to understand their respecitve role.
Suggestions: `cluster-leader` & `cluster-follower` / `cluster-publisher` & `cluster-subscriber` / something similar

Author: Thought about that as well, but it slightly contradicts a statement further up that a cluster doesn't have a designated subscriber/publisher role (on the cluster level), but can actually be both, depending on the table.
I changed the names and rephrased the paragraph above, to indicate that this is a simplification.

  CREATE TABLE doc.temperature_data (
    ts TIMESTAMP NOT NULL,
    temperature FLOAT NOT NULL
  );

Next, a publication is created. The publication marks our table as being
available for replication, but otherwise does not imply any activity yet.

.. code-block:: sql

  CREATE PUBLICATION temperature_publication FOR TABLE doc.temperature_data;

To verify the publication was created successfully, we query the
``pg_publication`` system table. It should contain one row with the publication
just added.

.. code-block:: sql

  SELECT *
  FROM pg_publication;

With this, we are already all set on the publication side.

Setting up a subscription
=========================

.. NOTE::

  All SQL statements in this section are executed on ``cluster-2``.

A subscription consists of connection information to the publishing cluster as
well as the name of the previously created publication to subscribe to.

Using the ``mode`` parameter, the connection can be established via either the
transport protocol or the PostgreSQL protocol. By setting the parameter to
``pg_tunnel``, we use the PostgreSQL protocol.

For a full overview of available parameters, please see the CREATE SUBSCRIPTION

> reference.

Reviewer: ToDo: Add link once docs are available

.. _logical_replication_setup:

==================================================
Logical replication setup between CrateDB clusters
==================================================

Logical replication is a method of data replication between multiple clusters.
As a publish/subscribe model, it allows a publishing cluster to make certain
tables available for subscription. Subscribing clusters pull changes from a
publication and replay them on their side.

.. rubric:: Table of contents

.. contents::
   :local:

.. _requirements:

Requirements
============
.. NOTE::

  Logical replication is available in CrateDB 4.8 and later.

To replicate data from one cluster to another, we need two CrateDB clusters.
For this example, we spin them up locally as two single-node Docker
containers.

> There exists no fixed designation of a cluster being a subscriber or publisher.

Reviewer 2: ```suggestion
There is no fixed designation of a cluster being a publisher or subscriber.
```

> One cluster can even take both roles for different tables. For the sake of
> simplicity, we will however refer here to the two clusters as
> ``cluster-publishing`` and ``cluster-subscribing``.

Reviewer 2: ```suggestion
One cluster can even take both roles for different tables. However, for the
sake of simplicity, we will refer to the two clusters as
``cluster-publishing`` and ``cluster-subscribing`` here.
```

.. code-block:: console

  sh$ docker run \
        --name cluster-publishing \
        --detach \
        --publish 4201:4200 \
        --publish 5433:5432 \
        crate/crate:nightly-4.8.0-20220401 \
        -Cdiscovery.type=single-node
  sh$ docker run \
        --name cluster-subscribing \
        --detach \
        --publish 4202:4200 \
        --publish 5434:5432 \
        crate/crate:nightly-4.8.0-20220401 \
        -Cdiscovery.type=single-node

We apply an offset to the ports of each cluster, so they don't overlap. The
Admin UIs of ``cluster-publishing`` and ``cluster-subscribing`` are accessible at
http://localhost:4201/ and http://localhost:4202/ respectively. All SQL
statements discussed below can be executed via the corresponding Admin UI.

As the subscriber pulls changes from the publisher, the publisher needs to accept

> incoming network connections from the subscriber. In our example setup, this is
> automatically given.

Reviewer 2: ```suggestion
incoming network connections from the subscriber. In our example setup, this is
automatically given because both containers are running on the same host.
For reaching each other, they will use the special ``host.docker.internal`` address.
```

Setting up a publication
========================

.. NOTE::

  All SQL statements in this section are executed on ``cluster-publishing``.

Before setting up the replication, we create a simple table that is going to be
the subject of replication. A table with the same name must not exist yet on
``cluster-subscribing``.

.. code-block:: sql

  CREATE TABLE doc.temperature_data (
    ts TIMESTAMP NOT NULL,
    temperature FLOAT NOT NULL
  );

Next, a publication is created. The publication marks our table as being
available for replication, but otherwise does not imply any activity yet.

.. code-block:: sql

  CREATE PUBLICATION temperature_publication FOR TABLE doc.temperature_data;

To verify the publication was created successfully, we query the
``pg_publication`` system table. It should contain one row with the publication
just added.

.. code-block:: sql

  SELECT *
  FROM pg_publication;

With this, we are already all set on the publication side.

Setting up a subscription
=========================

.. NOTE::

  All SQL statements in this section are executed on ``cluster-subscribing``.

> A subscription consists of connection information to the publishing cluster as

Reviewer 2: ```suggestion
A subscription needs connection information to the publishing cluster as
```

well as the name of the previously created publication to subscribe to.

Using the ``mode`` parameter, the connection can be established via either the
transport protocol or the PostgreSQL protocol. By setting the parameter to
``pg_tunnel``, we use the PostgreSQL protocol.

For a full overview of available parameters, please see the CREATE SUBSCRIPTION
reference.

.. code-block:: sql

  CREATE SUBSCRIPTION temperature_subscription
  CONNECTION 'crate://host.docker.internal:5433?user=crate&mode=pg_tunnel'
  PUBLICATION temperature_publication;

.. NOTE::

  The ``password`` parameter is omitted, as local connections using the ``crate``
  user don't require one. If you are connecting remotely, provide the credentials

>   of a user with ``DQL``, ``DML``, and ``DDL`` privileges on published tables.

Reviewer 2: Is this even true?
```suggestion
  of a user with ``DQL``, ``DML``, and ``DDL`` privileges on published tables
  and add a corresponding ``password`` parameter to the database URI.
```

Author: Good point, actually `DQL` is sufficient.

After a few seconds, the table ``doc.temperature_data`` should appear on
``cluster-subscribing``. At this point, it is still empty as we didn't insert
any data yet.

To verify the operational status of the subscription, the system tables

> pg_subscription and pg_subscription_rel can be queried. The below query returns

Reviewer 2: ```suggestion
``pg_subscription`` and ``pg_subscription_rel`` can be queried. The below query returns
```

Author: Those are going to become links to the reference documentation, but it's not possible yet, as cross-references are resolved against `latest` and those tables are only available in `master`. There are a few other missing links as well, hence this is still a draft PR.

> the name of the subscription (``subname``), its state (``srsubstate``) as well
> as any potential error message (``srsubstate_reason``):

Reviewer 2: ```suggestion
the name of the subscription (``subname``), its state (``srsubstate``), as well
as any potential error message (``srsubstate_reason``).
```

Using the ``mode`` parameter, the connection can be established via either the
transport protocol or the PostgreSQL protocol. By setting the parameter to
``pg_tunnel``, we use the PostgreSQL protocol.

For a full overview of available parameters, please see the CREATE SUBSCRIPTION
reference.

.. code-block:: sql

  CREATE SUBSCRIPTION temperature_subscription
  CONNECTION 'crate://host.docker.internal:5433?user=crate&mode=pg_tunnel'
  PUBLICATION temperature_publication;

.. NOTE::

  The ``password`` parameter is omitted, as local connections using the ``crate``
  user don't require one. If you are connecting remotely, provide ``user`` and
  ``password`` of a user with ``DQL`` privileges on published tables.

After a few seconds, the table ``doc.temperature_data`` should appear on
``cluster-subscribing``. At this point, it is still empty as we didn't insert
any data yet.

To verify the operational status of the subscription, the system tables
pg_subscription and pg_subscription_rel can be queried. The below query returns
the name of the subscription (``subname``), its state (``srsubstate``), as well
as any potential error message (``srsubstate_reason``).

.. code-block:: sql

>   SELECT subname, srsubstate, srsubstate_reason

Reviewer 3: r.srrelid::text gets a table name - might be useful to see the table name as status is granular per table.

Author: Thanks, added. I hope we can get some additional status views/tables in `sys` in the future (and integrate them into the Admin UI), then we can adjust this part of the article. I found the `pg_catalog` tables to be quite complicated.

>   SELECT subname, srsubstate, srsubstate_reason

Reviewer 3: r.srrelid::text gets a table name - might be useful to see the table name as status is granular per table.
